drivers/video/video_stm32_venc.c

```diff
@@ -0,0 +1,1026 @@
+#include <zephyr/dt-bindings/memory-attr/memory-attr.h>
+#include <zephyr/dt-bindings/memory-attr/memory-attr-sw.h>
+
+LOG_MODULE_REGISTER(video_stm32_venc, CONFIG_VIDEO_LOG_LEVEL);
+
+#define EWL_HEAP_ALIGNED_ALLOC(size)				\
+	mem_attr_heap_aligned_alloc((ATTR_SW_ALLOC_NON_CACHE <<	\
+				    DT_MEM_SW_ATTR_SHIFT), ALIGNMENT_INCR, size)
+#define EWL_HEAP_ALIGNED_FREE(block) mem_attr_heap_free(block)
+
+#define VENC_RATECTRL_FPS 30
+#define VENC_RATECTRL_QP 25
+
+#define TRACE_EWL
+
+#define EWL_USE_POLLING_SYNC
+
+#define EWL_TIMEOUT 100UL
+
+#define ALIGNMENT_INCR 8UL
+
+#define MEM_CHUNKS      32
+
+#ifdef TRACE_EWL
+u32 mem_counter = 0;
+#endif /* TRACE_EWL */
+
+typedef struct
+{
+  u32 clientType;
+  /* tables required for buffer alignment */
+  u32 *chunks[MEM_CHUNKS];
+  u32 *alignedChunks[MEM_CHUNKS];
+  u32 totalChunks;
+
+  struct k_sem complete;
+  i32 status;
+  i32 irq_status;
+  i32 irq_cnt;
+} VENC_EWL_TypeDef;
+
+#define ALIGNED_SIZE(x) ((((x) + ALIGNMENT_INCR - 1UL) / ALIGNMENT_INCR) * ALIGNMENT_INCR)
+
+static VENC_EWL_TypeDef ewl_instance;
+
+typedef void (*irq_config_func_t)(const struct device *dev);
+
+struct video_stm32_venc_data {
+	const struct device *dev;
+	struct video_format fmt;
+	struct k_fifo fifo_input;
+	struct k_fifo fifo_output_in;
+	struct k_fifo fifo_output_out;
+	uint32_t pixelformat;
+	uint32_t height;
+	uint32_t width;
+	uint32_t pitch;
+	struct video_buffer *vbuf;
+	H264EncInst encoder;
+	uint32_t frame_nb;
+	bool resync;
+};
+
+struct video_stm32_venc_config {
+	struct stm32_pclken pclken;
+	irq_config_func_t irq_config;
+};
+
+static int encoder_prepare(struct video_stm32_venc_data *data);
+static int encode_frame(struct video_stm32_venc_data *data);
+static int encoder_end(struct video_stm32_venc_data *data);
+static int encoder_start(struct video_stm32_venc_data *data);
+
+static inline H264EncPictureType to_h264pixfmt(uint32_t pixelformat)
+{
+	switch (pixelformat) {
+	case VIDEO_PIX_FMT_NV12:
+		return H264ENC_YUV420_SEMIPLANAR;
+	case VIDEO_PIX_FMT_RGB565:
+		return H264ENC_RGB565;
+	default:
+		return 0;
+	}
+}
+
+u32 EWLReadAsicID(void)
+{
+	LOG_DBG("%s>\n", __func__);
+	return LL_VENC_ReadRegister(0UL);
+	LOG_DBG("%s<\n", __func__);
+}
+
+EWLHwConfig_t EWLReadAsicConfig(void)
+{
+	EWLHwConfig_t cfg_info;
+	u32 cfgval;
+
+	LOG_DBG("%s>\n", __func__);
+
+	/* read first part of the configuration stored in register 63 */
+	cfgval = LL_VENC_ReadRegister(63UL);
+
+	cfg_info.maxEncodedWidth = cfgval & ((1U << 12U) - 1U);
+	cfg_info.h264Enabled = (cfgval >> 27U) & 1U;
+	cfg_info.vp8Enabled = (cfgval >> 26U) & 1U;
+	cfg_info.jpegEnabled = (cfgval >> 25U) & 1U;
+	cfg_info.vsEnabled = (cfgval >> 24U) & 1U;
+	cfg_info.rgbEnabled = (cfgval >> 28U) & 1U;
+	cfg_info.searchAreaSmall = (cfgval >> 29U) & 1U;
+	cfg_info.scalingEnabled = (cfgval >> 30U) & 1U;
+
+	cfg_info.busType = (cfgval >> 20U) & 15U;
+	cfg_info.synthesisLanguage = (cfgval >> 16U) & 15U;
+	cfg_info.busWidth = (cfgval >> 12U) & 15U;
```

Comment:
Nit: IMO, prefer hex notation (or use `GENMASK`) for masks with more than 1 bit.
```suggestion
	cfg_info.busType = (cfgval >> 20U) & 0xFU;
	cfg_info.synthesisLanguage = (cfgval >> 16U) & 0xFU;
	cfg_info.busWidth = (cfgval >> 12U) & 0xFU;
```